flink-clients/src/main/java/org/apache/flink/client/deployment/AbstractSessionClusterExecutor.java

```diff
@@ -50,18 +50,20 @@ public AbstractSessionClusterExecutor(@Nonnull final ClientFactory clusterClient
 	}
 
 	@Override
-	public CompletableFuture<JobClient> execute(@Nonnull final Pipeline pipeline, @Nonnull final Configuration configuration) throws Exception {
+	public CompletableFuture<? extends JobClient> execute(@Nonnull final Pipeline pipeline, @Nonnull final Configuration configuration) throws Exception {
 		final JobGraph jobGraph = ExecutorUtils.getJobGraph(pipeline, configuration);
 
 		try (final ClusterDescriptor<ClusterID> clusterDescriptor = clusterClientFactory.createClusterDescriptor(configuration)) {
 			final ClusterID clusterID = clusterClientFactory.getClusterId(configuration);
 			checkState(clusterID != null);
 
 			final ClusterClient<ClusterID> clusterClient = clusterDescriptor.retrieve(clusterID);
-			return clusterClient
-					.submitJob(jobGraph)
-					.thenApply(JobSubmissionResult::getJobID)
-					.thenApply(jobID -> new ClusterClientJobClientAdapter<>(clusterClient, jobID, false));
+			return clusterClient.submitJob(jobGraph).thenApply(jobClient -> {
+				if (jobClient instanceof ClusterClientJobClientAdapter) {
```

Comment:
This is pretty fragile and also not the best in terms of coding practices.

Actually after checking both versions of the PR, maybe it is better to keep the newly introduced `addOnCloseActions()` for the shutdown hook in the attached, per-job cluster case, and also keep the `moveOwnership` flag in the constructor of the `ClusterClientJobClientAdapter` and the `ClusterClient.submitJob()` method for closing, or not, the underlying cluster client. So, the `sae` flag will add a "closing action" and the `submitJob` will have an additional parameter.

In the future, we can think of how to get rid of it in a proper way.

I know that this is a bit of back and forth on my side, but I hope that it is not a big problem for you.

WDYT?

Reply:
```
I think I get your points here and push a commit 7d98e670ac6e917eb2c121493ba0b5925240899d dedicate for this comment.

Let sync for opinions.

In my understanding the idea is that we should keep initializing closing actions as near as the job client instanced. I agree that it is a bit tricky because 

(... original thoughts)
```

I have another idea for this stuff because we don't need constructing JobClient just in-place within `ClusterClient#submitJob` at all(instead, we return JobID there, and construct JobClient when we actually need one, thus we are able to rewrite close logic on consturaction). I'm going to a meeting and possibly back to push a commit reflect my thought hours later.

If we're still unhappy with the implementation I think it would be good we have a video conversation for consensus later.

Reply:
I am looking forward to your new commit later and then I will comment on that! Thanks for the effort.